Fix unresponsive unit buttons

esbudylin opened the pull request:
Prior to this commit, some of the unit buttons (e.g., Build City, Build Road) could become unresponsive. This issue was related to how the game handled button presses. When a button was pressed, the game emulated the press of an action hotkey, which should have been handled by the Game class. However, in some cases, the input simulation didn't work properly, and the game didn't react to the button presses.

To solve this issue, I decoupled the handling of button presses from the handling of hotkey input. Now, instead of imitating input, a unit button sends a signal to the Game class. In the Game class, the logic for handling action requests has been moved into a new method, which is now used for both handling signals from unit buttons and processing hotkey input.

esbudylin committed:
Fix unresponsive unit buttons

Prior to this commit, some of the unit buttons (e.g., Build City,
Build Road) could become unresponsive. This issue was related to how
the game handled button presses. When a button was pressed, the game
emulated the press of an action hotkey, which should have been handled
by the Game class. However, in some cases, the input simulation didn't work
properly, and the game didn't react to the button presses.

To solve this issue, I decoupled the handling of button presses from
the handling of hotkey input. Now, instead of imitating input, a unit
button sends a signal to the Game class. In the Game class, the logic for
handling action requests has been moved into a new method, which is
now used for both handling signals from unit buttons and processing
hotkey input.

pcen commented:
Is this a bug in Godot or incorrect usage? Conceptually this makes sense regardless

Sean-Brown commented on C7/Game.cs:
```diff
@@ -504,21 +512,21 @@ private void processActions() {
 			TileDirection dir = TileDirection.NORTH;
 			bool moveUnit = true;
-			if (Input.IsActionJustPressed(C7Action.MoveUnitSouthwest)) {
+			if (currentAction == C7Action.MoveUnitSouthwest) {
```
possibly should be a switch statement?

esbudylin committed:
Refactor: C7Action to TileDirection conversion

- Refactored the conversion from C7Action to TileDirection using a
switch expression. Moved the logic into the new method.
- Transformed fields of the C7Action class into constants to enable
their usage in the switch expression.

esbudylin committed:
Refactor: rewrite conversion method using nullable type

WildWeazel commented:
This seems sensible to me. I'm a little unsure about handling overlapping inputs for different actions (eg, if we implement Ctrl-G for the grid, which also activates G for go-to) but it's no worse than before.
